diff --git a/encoder/libav_encoder.cpp b/encoder/libav_encoder.cpp
index e4bcd90..9d5eebd 100644
--- a/encoder/libav_encoder.cpp
+++ b/encoder/libav_encoder.cpp
@@ -15,6 +15,7 @@
 #include <linux/videodev2.h>
 
 #include <chrono>
+#include <csignal>
 #include <iostream>
 
 #include "libav_encoder.hpp"
@@ -517,9 +518,17 @@ void LibAvEncoder::encode(AVPacket *pkt, unsigned int stream_id)
 		ret = av_interleaved_write_frame(out_fmt_ctx_, pkt);
 		if (ret < 0)
 		{
-			char err[AV_ERROR_MAX_STRING_SIZE];
-			av_strerror(ret, err, sizeof(err));
-			throw std::runtime_error("libav: error writing output: " + std::string(err));
+			// This means the connection suddenly dropped, handle gracefully
+			if (ret == -104 || ret == -32)
+			{
+				std::raise(SIGPIPE);
+			}
+			else
+			{
+				char err[AV_ERROR_MAX_STRING_SIZE];
+				av_strerror(ret, err, sizeof(err));
+				throw std::runtime_error("libav: error writing output: " + std::string(err));
+			}
 		}
 	}
 }
diff --git a/output/net_output.cpp b/output/net_output.cpp
index 6c3fe93..26e07bb 100644
--- a/output/net_output.cpp
+++ b/output/net_output.cpp
@@ -8,6 +8,9 @@
 #include <arpa/inet.h>
 #include <sys/socket.h>
 
+#include <csignal>
+#include <cerrno>
+
 #include "net_output.hpp"
 
 NetOutput::NetOutput(VideoOptions const *options) : Output(options)
@@ -105,8 +108,18 @@ void NetOutput::outputBuffer(void *mem, size_t size, int64_t /*timestamp_us*/, u
 	for (uint8_t *ptr = (uint8_t *)mem; size;)
 	{
 		size_t bytes_to_send = std::min(size, max_size);
+
 		if (sendto(fd_, ptr, bytes_to_send, 0, saddr_ptr_, sockaddr_in_size_) < 0)
-			throw std::runtime_error("failed to send data on socket");
+		{
+			if (errno != EPIPE)
+			{
+				if (errno == ECONNRESET)
+					std::raise(SIGPIPE);
+				else
+					throw std::runtime_error("failed to send data on socket");
+			}
+		}
+
 		ptr += bytes_to_send;
 		size -= bytes_to_send;
 	}
